homeassistant/components/tag/trigger.py

```diff
@@ -1,37 +1,61 @@
 """Support for tag triggers."""
 import voluptuous as vol
 
-from homeassistant.components.homeassistant.triggers import event as event_trigger
 from homeassistant.const import CONF_PLATFORM
+from homeassistant.core import HassJob, callback
 from homeassistant.helpers import config_validation as cv
 
 from .const import DEVICE_ID, DOMAIN, EVENT_TAG_SCANNED, TAG_ID
 
 TRIGGER_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_PLATFORM): DOMAIN,
-        vol.Required(TAG_ID): cv.string,
-        vol.Optional(DEVICE_ID): cv.string,
+        vol.Required(TAG_ID): vol.All(cv.ensure_list, [cv.string]),
+        vol.Optional(DEVICE_ID): vol.All(cv.ensure_list, [cv.string]),
     }
 )
 
 
 async def async_attach_trigger(hass, config, action, automation_info):
     """Listen for tag_scanned events based on configuration."""
-    tag_id = config.get(TAG_ID)
-    device_id = config.get(DEVICE_ID)
-    event_data = {TAG_ID: tag_id}
-
-    if device_id:
-        event_data[DEVICE_ID] = device_id
-
-    event_config = {
-        event_trigger.CONF_PLATFORM: "event",
-        event_trigger.CONF_EVENT_TYPE: EVENT_TAG_SCANNED,
-        event_trigger.CONF_EVENT_DATA: event_data,
-    }
-    event_config = event_trigger.TRIGGER_SCHEMA(event_config)
-
-    return await event_trigger.async_attach_trigger(
-        hass, event_config, action, automation_info, platform_type=DOMAIN
+    tag_ids = config.get(TAG_ID)
+    device_ids = config.get(DEVICE_ID)
+
+    tag_data_schema = vol.Schema(
+        {
+            vol.Required(TAG_ID): vol.In(tag_ids),
+        },
+        extra=vol.ALLOW_EXTRA,
     )
+
+    if device_ids:
+        tag_data_schema = tag_data_schema.extend(
+            {
+                vol.Required(DEVICE_ID): vol.In(device_ids),
+            }
+        )
+
+    job = HassJob(action)
+
+    @callback
+    def handle_event(event):
+        """Listen for tag scan events and calls the action when data matches."""
+        try:
+            tag_data_schema(event.data)
+        except vol.Invalid:
+            # If event doesn't match, skip tag event
+            return
```

Comment:
A voluptuous schema is possible but I think it's cleaner to just check the data directly. 

```suggestion
        if event.data.get(TAG_ID) not in device_ids or (device_ids and event.data.get(DEVICE_ID) not in device_ids):
            return
```